### DISCUSSION TITLE
Auxkernels and AD

### DISCUSSION THREAD

**Discussion by joe61vette** [idaholab/moose#27544] [Q&A General] [answered]
I have a compound auxvariable (mdot) that is the product of density, velocity and area. I use an aux instead of a material property due to a "product of averages is not equal to the average of the products" type of problem when variables are interpolated to QPs. I would like to use the AD system but it seems like auxkernels only return Real values. Is it possible for an auxvariable to be of type ADReal? I haven't found an example yet, so I am guessing not.
Thanks,
  Joe Kelly

**Comment by GiudGiud** [comment]
Hi Joe

No it's not possible for aux variables to be AD.
So you want to avoid evaluating the material property on the face by evaluating each component? You want to make sure face values are computed from the average of the values of that material property on the elements?
With Functor material properties we can define how the property gets evaluated for each argument (element, face, point, etc)

Guillaume

**Reply by joe61vette** [reply]
Thanks Guillaume. I'll take a look at functor properties.  I have only used them in an FVM context and this is FEM (1D).  After I read up on it, I'll either mark this as answered or have a better defined question. Below is an excerpt that explains the problem a bit better.

"To simplify the discussion, we’ll first consider a constant density case. It is essential to note that in this implementation the values of the flow area and the velocity variables only exist for the nodal points. Within a kernel, the values for the quadrature points are interpolated from the actual nodal values according to the type of quadrature being used. Again, to keep this simple, we’ll consider the most elementary type of quadrature: first-order Gauss. Thus, there is only one quadrature point located at the midpoint of the element and linear interpolation is used. Therefore,

A[0] = (1/2)*(A0 + A1) 
U[0] = (1/2)* (U0 + U1)

Let’s consider the case where the flow area of an element doubles over its length and the velocity and area at the bottom of the element both have a value of unity. The corresponding volumetric flux would then have a value of 1 m3/s. For a constant density case, the volumetric flux at the top of the element would also have the same value. However, using the quadrature point values we would have

A[0] = (1/2)* (1+2) = 1.5 
U[0]= (1/2)* (1+0.5)=0.75

so that the product U · A is equal to 1.125 rather than unity. Consequently, using the product of the quadrature values for velocity and flow area within the convective term can result in large mass errors rendering the calculation meaningless."

**Reply by joe61vette** [reply]
After looking at the functor material description, I don't believe that it solves the problem described above. Namely, for a 1D problem, only the velocity calculated at the nodal points is guaranteed to conserve the mass flow rate. If you interpolate each quantity (rho, velocity and area) to the QPs in forming a residual, that product does not equal the mass flow rate.

**Reply by GiudGiud** [reply, marked as answer]
Well you're right that functor material properties were initially meant for facilitating on-the-fly evaluations and therefore meant it would be easy to combine functors / call them anywhere. So if mdot = rho v A then they allow to compute mdot(anywhere = x) as rho(x) * v(x) * A(x)

but it does not have to be that way. If I define a functor mdot that is equal to rho v A only on the nodes (definition of the functor for the NodalArg spatial argument), and equal to an interpolation of mdot between nodes elsewhere (definition of the functor for other spatial arguments, for example ElemQpArg, a Qp inside an element), then it does what you want. 

You can look at ConstantFunctor in MooseFunctor.h for one of the simplest functor definitions.
Then PiecewiseByBlockLambdaFunctor in its own header for a more complex one.
My idea would be to create the one you need.
@lindsayad is the most knowledgeable on functors

Then for the user to create them, that will have to be done through a FunctorMaterial, which will create the functor as a functor material property.

**Reply by joe61vette** [reply]
Thanks Guillaume.  I think I understand and will give it a try.

**Reply by GiudGiud** [reply]
sounds good. There is also a functor that leverages a map by faces. that can be an inspiration too. It's in Navier Stokes module
src/utils/FaceCenteredMapFunctor.C

**Reply by lindsayad** [reply]
There's a flaw in the thought process here. There is no interpolation in finite elements, just evaluation of the approximate solution. There is nothing special about the solution evaluation at the nodes vs. the interior of the elements. The evaluation of U at an interior element quadrature point is just as correct or incorrect as its evaluation at a node. Same with A. So you should not regard the product of U and A and rho at nodes as something special upon which you should base all other evaluations. If you have an issue with evaluation of mdot because of its composite nature, the fix would be to cast it as a degree of freedom in your system of equations, not try to project an auxiliary mdot into a finite element space
